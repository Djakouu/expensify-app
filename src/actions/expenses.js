import uuid from 'uuid'
import database from '../firebase/firebase'

export const addExpense = expense => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return dispatch => {
        const {
            description = '',
            amount = 0,
            note = '', 
            createdAt = 0 
        } = expenseData;
        const expense = { description, amount, note, createdAt }
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense =({ id } = {}) => {
    return dispatch => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        })
    }
};

// EDIT EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    updates,
    id
});

export const startEditExpense = (id, updates) => {
    return dispatch => {
        return database.ref(`expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates))
        });
    }
};

// SET EXPENSE
export const setExpenses = expenses => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach(childSnapshot => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    })
};