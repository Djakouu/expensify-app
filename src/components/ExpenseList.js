import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = props => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.length === 0 ? (
            <p> No expenses </p>
        ) : (
            props.expenses.map(expense => {
                return <ExpenseListItem key={expense.id} {...expense} />
            })
        )}
        
    </div>
);

const mapStatetoProps = state => ({
        expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStatetoProps)(ExpenseList);
