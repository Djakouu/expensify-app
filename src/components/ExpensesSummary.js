import React from 'react';
import { connect } from 'react-redux'
import { Link } from 'react-router-dom';
import numeral from "numeral";
import '../adjustments/numeral-locale-fr'
import getVisibleExpenses from '../selectors/expenses';
import { getExpensesTotal } from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
    <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title" >Viewing <span>{props.expenses.length < 2 ? props.expenses.length : props.expenses.length}</span>
            {props.expenses.length < 2 ? ' expense ' : ' expenses '}
            totalling <span>{numeral(getExpensesTotal(props.expenses) / 100).format('0,0.00 $')}</span>
            </h1>
            <div className="page-header__actions">
                <Link className="button" to="/create"> Add Expense</Link>
            </div>
        </div>
    </div>
);

const mapStatetoProps = state => ({
    expenses: getVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStatetoProps)(ExpensesSummary);