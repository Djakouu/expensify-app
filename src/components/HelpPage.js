import React from 'react';

const Help = () => (
    <div>
        <p> This is from Help</p>
    </div>
);

export default Help;