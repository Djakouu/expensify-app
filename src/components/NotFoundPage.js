import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        <p>404 ­— <Link to="/">Go home</Link></p>
    </div>
);

export default NotFoundPage;