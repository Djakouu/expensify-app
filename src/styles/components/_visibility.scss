.show-for-mobile {
    @media (min-width: $desktop-breakpoint) {
        display: none;
    }
}

.show-for-desktop {
    @media (max-width: $desktop-breakpoint - .1rem) {
        display: none;
    }
}