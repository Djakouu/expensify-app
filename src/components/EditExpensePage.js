import React from 'react';
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpense extends React.Component {
    onSubmit = expense => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/')
    }
    onClick = () => {
        this.props.startRemoveExpense({id: this.props.expense.id})
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>
                    Remove
                </button>
            </div>
        )
    }
}

// const Edit = (props) => ( 
//     <div>
//         <h1>Edit Expense</h1>
//         <ExpenseForm 
//             expense={props.expense}
//             onSubmit={(expense) => {
//                 props.dispatch(editExpense(props.expense.id, expense));
//                 props.history.push('/')
//             }}
//         />
//         <button onClick={() => {
//             props.dispatch(removeExpense({id: props.expense.id}));
//             props.history.push('/')
//         }
//         }>
//             Remove
//         </button>
//     </div>
// );

const mapStateToProps = (state, props) => ({
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchProps = (dispatch, props) => ({
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpense: data => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchProps)(EditExpense);