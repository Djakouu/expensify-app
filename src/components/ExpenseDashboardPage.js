import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const Expense = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default Expense;