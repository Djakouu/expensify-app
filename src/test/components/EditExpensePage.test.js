import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';


let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpense 
        expense={expenses[1]} 
        startEditExpense={startEditExpense} 
        startRemoveExpense={startRemoveExpense} 
        history={history} 
        />);
});

test('should render startEditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle start edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('should handle start remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id: expenses[1].id});
});










