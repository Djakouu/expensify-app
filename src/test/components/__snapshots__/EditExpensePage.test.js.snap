// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render startEditExpensePage correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 200,
        "createdAt": -345600000,
        "description": "gas",
        "id": 2,
        "note": "Note 2",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
